### ISSUE TITLE
Map Subtyping Support

### ISSUE THREAD

**Issue by albsch** [etylizer/etylizer#30]
Subtyping for Erlang map constructs is currently being investigated and implemented by @erdemYl .

See the type spec for more information:

https://www.erlang.org/doc/reference_manual/typespec.html

**Comment by erdemYl** [comment]
...State Update... (informal)


Map types require an interpretation backed by Semantic Subtyping. The chosen interpretation will be the founding part of the subtyping algorithm. A decision has to be made on interpreting  #{t() => s()}. There were two choices:

    (1) only allow keys of type t()
    (2) #{t() => s()}  =  #{t() => s(), _=>_}
  
  
Remember that the syntax  #{t() => s(), ...}  is absent in erlang, therefore some programmers tend to think in the way of (2). 


We tried to find out a hint that could justify "some" as "many". We analyzed map usage from 10 projects that use type specs and maps in reasonable amounts. The result is that  _=>_  appears approx. 24% in specs. Thus, a remarkable amount of people "make their intention explicit" that map may contain other keys than t(). For the rest 76% no clear comments could be made whether they adhere to (2) or not.


Moreover, functional programming such as  keys(#{Key => _}) -> [Key]  could be endangered by (2), since [Key] list now contains hidden types.


Because of this reasons and that (2) seems to demand contravariant key types, which also poses a danger for functional programming, we decided to go for the (1).


Subtyping theory of maps will either be the one defined by us, or the one that is newly (and still) developed by Giuseppe Castagna (see https://icfp23.sigplan.org/details/icfp-2023-papers/8/Typing-Records-Maps-and-Structs).
